File: 12월 4주차/징검다리.py
def go(m, arr, n):
    start = 0
    for i in range(len(arr)):
        cur = arr[i]
        if cur - start >= m:
            start = cur
        else:
            n -= 1
            if n < 0:
                return True
    return False

def solution(distance, rocks, n):
    answer = 0
    rocks.sort()
    rocks += [distance]
    l,r = 0, distance
    while l <= r:
        m = (l+r)//2
        if go(m, rocks, n):
            r = m - 1
        else: l = m + 1 
    return r

d,r,n = 25, [2, 14, 11, 21, 17], 2
solution(d,r,n)

File: 1월 3주차/징검다리.js
function go(m,n,rocks){
    var l = rocks.length
    var cur = 0
    for (var i=0; i<l; i++){
        if (rocks[i]-cur < m){
            n -= 1
            if (n < 0){
                return false
            }
        } else{
            cur = rocks[i]
        }
    }
    return true
}

function solution(distance, rocks, n) {
    rocks.sort((a,b)=>{
        return a-b
    })
    
    var l,r,m
    l = 0
    r = distance
    while (l<=r){
        m = parseInt((l+r)/2)
        if (go(m,n,rocks)){
            l = m+1
        } else {
            r = m-1
        }
    }
    
    return r;
}